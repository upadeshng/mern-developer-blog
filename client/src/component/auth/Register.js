import React, { useState } from 'react';
import { Link, withRouter, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';

import { Button } from 'react-bootstrap';

import { register } from '../../redux/action/auth';
import { setAlert } from '../../redux/action/alert';
import PropTypes from 'prop-types';

import DisplayValidation from '../DisplayValidation';

const Register = ({ setAlert, register, history, isAuthenticated }) => {
  const [formData, setFormData] = useState({
    name: '',
    email: '',
    password: '',
    password2: '',
  });

  const { name, email, password, password2 } = formData;

  const onChange = (e) =>
    setFormData({ ...formData, [e.target.name]: e.target.value });

  const onSubmit = (e) => {
    e.preventDefault();

    if (password !== password2) {
      setAlert('password do not match', 'danger');
    } else {
      register({ name, email, password }, history);
    }
  };

  // Redirect to Dashboard after register
  if (isAuthenticated) {
    return <Redirect to='/dashboard' />;
  }

  return (
    <>
      {/* <DisplayValidation validationErrors={this.props.validationData} /> */}

      <h1 className='large text-primary'>Sign Up</h1>
      <p className='lead'>
        <i className='fas fa-user'></i> Create Your Account
      </p>
      <form
        className='form'
        action='create-profile.html'
        onSubmit={(e) => onSubmit(e)}
      >
        <div className='form-group'>
          <input
            type='text'
            placeholder='Name'
            name='name'
            value={name}
            onChange={(e) => onChange(e)}
          />
        </div>
        <div className='form-group'>
          <input
            type='email'
            placeholder='Email Address'
            name='email'
            value={email}
            onChange={(e) => onChange(e)}
          />
          <small className='form-text'>
            This site uses Gravatar so if you want a profile image, use a
            Gravatar email
          </small>
        </div>
        <div className='form-group'>
          <input
            type='password'
            placeholder='Password'
            name='password'
            value={password}
            minLength='6'
            onChange={(e) => onChange(e)}
          />
        </div>
        <div className='form-group'>
          <input
            type='password'
            placeholder='Confirm Password'
            name='password2'
            value={password2}
            minLength='6'
            onChange={(e) => onChange(e)}
          />
        </div>

        <Button
          className='btn btn-primary submit-btn'
          onClick={(e) => onSubmit(e)}
        >
          Register
        </Button>
      </form>
      <p className='my-1'>
        Already have an account?
        <Link to='/login' className='login-link'>
          Sign In
        </Link>
      </p>
    </>
  );
};

Register.propTypes = {
  setAlert: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool,
};
const mapStateToProps = (state) => ({
  validationData: state.authReducer.validationData,
  isAuthenticated: state.authReducer.isAuthenticated,
});

export default connect(mapStateToProps, { register, setAlert })(
  withRouter(Register)
);
