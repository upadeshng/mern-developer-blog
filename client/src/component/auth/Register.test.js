import React from 'react';
import Register from './Register';
import Action from './action';

import { BrowserRouter } from 'react-router-dom';

import { Form } from 'react-bootstrap';

import { shallow, mount, render } from 'enzyme';
import { Provider, connect } from 'react-redux';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import axios from 'axios';
import moxios from 'moxios';

import { mapStateToProps, mapDispatchToProps } from './Register';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const setup = () => {
  //   const wrapper = shallow(<Register />);
  let initialState = {
    apollo: { apolloLink: 'test value goes here' },
    signupReducer: { validationData: null },
  };
  let store = mockStore(initialState);

  let props = {
    validationData: { errors: null },
    authActions: {
      login: jest.fn(),
      resetPassword: jest.fn(),
    },
  };

  // let resetWrapper = shallow(<Register store={store} {...props} />);

  let wrapper = mount(
    <BrowserRouter>
      <Provider store={store} {...props}>
        <Register />
      </Provider>
    </BrowserRouter>
  );

  return { store, props, wrapper };
};

describe('Register', () => {
  //   const wrapper = shallow(<Register />);
  let initialState = { apollo: { apolloLink: 'test value goes here' } };
  let store = mockStore(initialState);

  const wrapperSnapshot = shallow(
    <Provider store={store}>
      {}
      // Provides the store to your App
      <Register />
    </Provider>
  );

  //   Snapshot
  it('Renders correctly with snapshot', () => {
    expect(wrapperSnapshot).toMatchSnapshot();
  });

  const { wrapper } = setup();
  let componentInstance = wrapper.find('Register').instance();

  /* let wrapper = mount(
    <BrowserRouter>
      <Provider store={store}>
        <Register />
      </Provider>
    </BrowserRouter>
  ); */

  //   Form element value
  const nameValue = 'Desh';
  const emailValue = 'upadesh.ng@gmail.com';
  const passwordValue = 'Secret123';
  const password2Value = 'Secret123';

  describe('Has form elements', () => {
    // Name input
    it('renders name input', () => {
      //   console.log(wrapper.debug({ verbose: true }));

      const input = wrapper.find('input').find({ name: 'name' });
      expect(input).toHaveLength(1);
    });

    // Email input
    it('renders email input', () => {
      const input = wrapper.find('input').find({ name: 'email' });
      expect(input).toHaveLength(1);
    });

    // Password input
    it('renders password', () => {
      const input = wrapper.find('input').find({ name: 'password' });
      expect(input).toHaveLength(1);
    });

    // Password2 input
    it('renders password2', () => {
      const input = wrapper.find('input').find({ name: 'password2' });
      expect(input).toHaveLength(1);
    });
  });

  //   render login link
  it('renders login link', () => {
    // console.log(wrapper.debug());
    const loginLink = wrapper.find('.login-link').at(0);
    expect(loginLink).toHaveLength(1);
  });

  //   input change
  describe('when typing in input elements', () => {
    // Updates name
    it('updates name input', () => {
      wrapper
        .find('input')
        .find({ name: 'name' })
        .simulate('change', { target: { value: nameValue } });

      //   console.log(wrapper.debug());
      expect(componentInstance.state.name).toEqual(nameValue);
    });

    // Updates email
    it('updates email input', () => {
      wrapper
        .find('input')
        .find({ name: 'email' })
        .simulate('change', { target: { value: emailValue } });

      expect(componentInstance.state.email).toEqual(emailValue);
    });

    // Updates password
    it('updates email input', () => {
      wrapper
        .find('input')
        .find({ name: 'password' })
        .simulate('change', { target: { value: passwordValue } });

      expect(componentInstance.state.password).toEqual(passwordValue);
    });

    // Updates password2
    it('updates email input', () => {
      wrapper
        .find('input')
        .find({ name: 'password2' })
        .simulate('change', { target: { value: password2Value } });

      expect(componentInstance.state.password2).toEqual(password2Value);
    });

    // Compares password & password2
    it('compares password & password2', () => {
      wrapper
        .find('input')
        .find({ name: 'password' })
        .simulate('change', { target: { value: passwordValue } });

      wrapper
        .find('input')
        .find({ name: 'password2' })
        .simulate('change', { target: { value: password2Value } });

      const passwordUpdated = componentInstance.state.password;
      const password2Updated = componentInstance.state.password2;
      expect(passwordUpdated).toEqual(password2Updated);
    });
  });

  describe('when submitted', () => {
    const testValues = {
      name: nameValue,
      email: emailValue,
      password: passwordValue,
      password2: password2Value,
      handleSubmit: jest.fn(),
    };

    /* it('has `handlesubmit()` function', () => {
      const { store, props } = setup();
      let wrapper1 = mount(
        <BrowserRouter>
          <Provider store={store} {...props}>
            <Register {...testValues} />
          </Provider>
        </BrowserRouter>
      );

      wrapper1.instance().handleSubmit_();
    }); */

    it('Submit works', () => {
      const { store, props } = setup();
      let wrapper1 = mount(
        <BrowserRouter>
          <Provider store={store} {...props}>
            <Register {...testValues} />
          </Provider>
        </BrowserRouter>
      );

      wrapper1.find('button').simulate('click');
    });
  });

  describe('selectAvatar', () => {
    beforeEach(() => {
      // Runs before each test in the suite
      const store = mockStore();
      store.clearActions();
    });

    it('should set uploading back to false on file successful upload', () => {
      const data = 'dummydata';
      moxios.stubRequest('/api/users', {
        status: 200,
        response: data,
      });
      const props = {
        onFileNameChange: jest.fn(),
      };

      // let componentInstance = wrapper.find('Register').instance();

      const { wrapper } = setup();
      let componentInstance = wrapper.find('Register').instance();

      componentInstance.props.postSignup('dummydata');
      wrapper.update();

      const expectedActions = [
        {
          payload: 1,
          type: 'select_avatar',
        },
      ];

      // store.dispatch({ type: 'DECREMENT' });
      // expect(store.getActions()).toEqual(expectedActions);

      // expect(props.onFileNameChange).toHaveBeenCalledWith(data);
      // expect(wrapper.state('uploading')).toBe(false);

      // const wrapper = shallow(<ImageUpload {...props} />);
      // wrapper.setState({ uploading: true });
      // wrapper.instance().sendFormData('dummydata');
      // wrapper.update();
      // expect(props.onFileNameChange).toHaveBeenCalledWith(data);
      // expect(wrapper.state('uploading')).toBe(false);
    });
  });

  //
});
