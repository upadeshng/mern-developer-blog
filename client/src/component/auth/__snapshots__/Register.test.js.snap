// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Renders correctly with snapshot 1`] = `ShallowWrapper {}`;
